framework/src/constraints/ConstraintWarehouse.C

```diff
@@ -321,13 +321,15 @@ ConstraintWarehouse::subdomainsCovered(std::set<SubdomainID> & subdomains_covere
     const auto & objects = pr.second.getActiveObjects();
     for (const auto & mc : objects)
     {
-      const MooseVariableFEBase * lm_var = &mc->variable();
+      const MooseVariableFEBase * lm_var = mc->variable_ptr();
       if (lm_var)
       {
         unique_variables.insert(lm_var->name());
         const std::set<SubdomainID> & lm_subdomains = lm_var->activeSubdomains();
         subdomains_covered.insert(lm_subdomains.begin(), lm_subdomains.end());
       }
+      else // May be a penalty constraint: Check in the constraint
+        subdomains_covered.insert(mc->secondarySubdomain());
```

Comment:
We should actually remove all the code before this last line. This is supposed to measure the subdomains covered by the constraint. Checking the LM-var subdomains is actually inaccurate. That is my fault from before. Fix this down in the displaced constraints too please.

Reply:
I wonder if we shouldn't check for kernel coverage for mortar constraints at all. Or, if we should check kernel coverage by questioning the object for whether an associated LM is expected (instead of what I had done earlier of automatically adding the subdomain if the constraint doesn't have an associated LM).

Reply:
Ok my recommendation was not good. 😞 

There are two things the `subdomainsCovered` method is supposed to do: register the blocks the residual objects cover and register the variables the residual objects act upon. The ultimate goal is to 
1. make sure that each mesh subdomain has a residual object acting on it
2. make sure that each variable has a residual object acting on it

So we need to keep the `unique_variables.insert` line when there is an `lm_var`, so we do indeed need a `variablePtr` method like you had before (note the MOOSE style; you had an underscore before). But for the `subdomains_covered` we should always insert with just the `mc->secondarySubdomain()` and `mc->primarySubdomain()`.

I'm sorry for leading you astray initially. That's on me!